### PR TITLE
Feat/via withdrawal client lib

### PR DESCRIPTION
**Pull request by 0xatomFusion**
## What ❔

Implemented the via withdrawal client.
- Fetch the pubdata from the da client.
- Decode the pubdata and select the withdrawal.
- Added a withdraw cmd to via cli.

PRs related #68 #70 
## Why ❔

This PR is required to list the withdrawals that were processed in an L1 batch.

## Test
1. Make via
2. Bridge BTC to L2 
```
via verifier deposit  --amount 100  --receiver-l2-address 0x36615Cf349d7F6344891B1e7CA7C72883F5dc049
```
3. Withdraw token from L2 -> L1
```
via token withdraw   --amount 5   --receiver-l1-address bcrt1qx2lk0unukm80qmepjp49hwf9z6xnz0s73k9j56
```
4. List the withdrawals in the block 2
```
cargo run --example withdraw 2
```
You should see your withdrawal
Screenshot from 2024-12-04 18-20-54

5. done
## Checklist

- [X] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [X] Tests for the changes have been added / updated.
- [] Documentation comments have been added / updated.
- [X] Code has been formatted via `zk fmt` and `zk lint`.

### CODE REVIEW

File: core/lib/via_withdrawal_client/src/client.rs [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+    pub async fn get_withdrawals(&self, blob_id: &str) -> anyhow::Result<Vec<WithdrawalRequest>> {
+        let pubdata_bytes = self._fetch_pubdata(blob_id).await?;
+        let pubdata = Pubdata::decode_pubdata(pubdata_bytes)?;
+        let l2_bridge_metadata = WithdrawalClient::_list_l2_bridge_metadata(&pubdata);
+        let withdrawals = WithdrawalClient::_get_valid_withdrawals(l2_bridge_metadata);
+        Ok(withdrawals)
+    }
+
+    async fn _fetch_pubdata(&self, blob_id: &str) -> anyhow::Result<Vec<u8>> {
```
Comment (steph-rs) [inline core/lib/via_withdrawal_client/src/client.rs:30]: maybe we can keep our previous rust-convention and don't use leading underscore for private functions? it seems weird for me, but if you insist, that's ok
https://doc.rust-lang.org/1.0.0/style/style/naming/README.html

File: core/lib/via_withdrawal_client/src/client.rs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        let l2_bridges_hash =
+            H256::from(H160::from_str(L2_BASE_TOKEN_SYSTEM_CONTRACT_ADDR).unwrap());
+        let l2_to_l1_messages_hashmap = WithdrawalClient::_l2_to_l1_messages_hashmap(&pubdata);
+        for log in pubdata.user_logs.clone() {
+            // Ignore the logs if not from emitted from the L2 bridge contract
+            if log.key != l2_bridges_hash {
+                continue;
+            };
+
+            if log.key != l2_bridges_hash {
```
Comment (steph-rs) [inline core/lib/via_withdrawal_client/src/client.rs:58]: repeated code

File: core/lib/via_withdrawal_client/src/client.rs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            let hash = H256::from(keccak256(&message));
+            hashes.insert(hash, message.clone());
+        }
+        hashes
+    }
+
+    fn _list_l2_bridge_metadata(pubdata: &Pubdata) -> Vec<L2BridgeLogMetadata> {
+        let mut withdrawals: Vec<L2BridgeLogMetadata> = Vec::new();
+        let l2_bridges_hash =
+            H256::from(H160::from_str(L2_BASE_TOKEN_SYSTEM_CONTRACT_ADDR).unwrap());
```
Comment (steph-rs) [inline core/lib/via_withdrawal_client/src/client.rs:50]: don't we need a config for that? system_conntract_addrs will be the same for any chain?

Comment (steph-rs) [reply core/lib/via_withdrawal_client/src/client.rs:50]: never mind, it seems ok

File: core/lib/via_withdrawal_client/src/types.rs [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+            tx_number_in_block,
+            sender,
+            key,
+            value,
+        })
+    }
+}
+
+/// Helper function to read a specific number of bytes
+fn _read_bytes<R: Read>(reader: &mut R, num_bytes: usize) -> anyhow::Result<Vec<u8>> {
```
Comment (steph-rs) [inline core/lib/via_withdrawal_client/src/types.rs:110]: same underscore

**Review by steph-rs — CHANGES_REQUESTED** [review]

File: core/lib/via_withdrawal_client/src/client.rs (hunk 1 above)
Comment (0xatomFusion) [reply core/lib/via_withdrawal_client/src/client.rs:30]: Fixed

**Review by steph-rs — APPROVED** [review]

**Review by mpavlovic-txfusion — APPROVED** [review]

**Review by irnb — APPROVED** [review]

**Pull request merged**